packages/cosmos/src/error.rs

```diff
@@ -248,30 +250,102 @@ pub enum QueryErrorDetails {
+    #[error("Cosmos SDK error code {error_code} returned: {status:?}")]
+    CosmosSdk {
+        error_code: u32,
+        status: tonic::Status,
+    },
+    #[error("Error parsing message into expected type: {0:?}")]
+    JsonParseError(tonic::Status),
+    #[error("{0:?}")]
+    FailedToExecute(tonic::Status),
+}
+
+pub(crate) enum QueryErrorCategory {
+    /// Should retry, kill the connection
+    NetworkIssue,
+    /// Don't retry, connection is fine
+    ConnectionIsFine,
+    /// No idea, make a test query and try again
+    Unsure,
 }
 
 impl QueryErrorDetails {
     /// Indicates that the error may be transient and deserves a retry.
-    pub(crate) fn should_be_retried(&self) -> bool {
+    pub(crate) fn error_category(&self) -> QueryErrorCategory {
+        use QueryErrorCategory::*;
         match self {
             // Not sure, so give it a retry
-            QueryErrorDetails::Unknown(_) => true,
+            QueryErrorDetails::Unknown(_) => Unsure,
             // Yup, may as well try to connect again.
-            QueryErrorDetails::ConnectionTimeout => true,
+            QueryErrorDetails::ConnectionTimeout => NetworkIssue,
             // Same here, maybe it was a bad connection.
-            QueryErrorDetails::QueryTimeout => true,
+            QueryErrorDetails::QueryTimeout => NetworkIssue,
             // Also possibly a bad connection
-            QueryErrorDetails::ConnectionError(_) => true,
-            QueryErrorDetails::NotFound(_) => false,
+            QueryErrorDetails::ConnectionError(_) => NetworkIssue,
+            QueryErrorDetails::NotFound(_) => ConnectionIsFine,
+            QueryErrorDetails::CosmosSdk { .. } => ConnectionIsFine,
+            QueryErrorDetails::JsonParseError(_) => ConnectionIsFine,
+            QueryErrorDetails::FailedToExecute(_) => ConnectionIsFine,
         }
     }
 
     pub(crate) fn from_tonic_status(err: tonic::Status) -> QueryErrorDetails {
         // For some reason, it looks like Osmosis testnet isn't returning a NotFound. Ugly workaround...
         if err.message().contains("not found") || err.code() == tonic::Code::NotFound {
-            QueryErrorDetails::NotFound(err.message().to_owned())
-        } else {
-            QueryErrorDetails::Unknown(err)
+            return QueryErrorDetails::NotFound(err.message().to_owned());
+        }
+
+        if let Some(error_code) = extract_cosmos_sdk_error_code(err.message()) {
+            return QueryErrorDetails::CosmosSdk {
+                error_code,
+                status: err,
+            };
         }
+
+        if err.message().starts_with("Error parsing into type ") {
+            return QueryErrorDetails::JsonParseError(err);
+        }
+
+        if err.message().starts_with("failed to execute message;") {
+            return QueryErrorDetails::FailedToExecute(err);
+        }
+
+        QueryErrorDetails::Unknown(err)
+    }
+}
+
+fn extract_cosmos_sdk_error_code(message: &str) -> Option<u32> {
```

Comment:
Wish it provided something better. Do you think it sets the code here properly now: https://docs.rs/tonic/latest/tonic/struct.Status.html#method.code ?

Reply:
I wish that too. I just did a bunch of testing, and I never see useful codes unfortunately :(